### ISSUE TITLE
Inaccurate gradient estimation fo log penalization

### ISSUE THREAD

**Issue by AxelBreuer** [cvxgrp/cvxpylayers#135]
Hi,
I have a simple trading stategy (trading 1 stock, max buy/sell =1.0, max position =10.).
I supposed that I know all future prices.
The objective function I maximize is equal to:
PnL(trading) - lambda*log_penalization(when trading close to 1.0 or -1.0).
I want to calculate the sensitivity of my optimal trading strategy at time=1 with respect to lambda.
The gradient calculated with cvxpylayers (-45.69) is very different from its finite difference approximation (-2.10).
I have tried many different tunings of the numerical engine behind cvxpylayer (including interfacing cvpylayers/diffcp with ECOS and adding an extra mode="lsmr" see https://github.com/cvxpy/cvxpy/issues/1915) nothing reduced the difference between both sensitivities.
A bit like mentionned in https://github.com/cvxgrp/cvxpylayers/issues/85, I am wondering whether this difference stems from an internal bug in cvxpylayers or if its a numerical instability due to log ?

Below the code of the trading strategy:

```python
import cvxpy as cp
import numpy as np
import torch
from cvxpylayers.torch import CvxpyLayer

"""
generate fake daily close prices
"""

D = 10  # trading horizon in number of days

rng = np.random.default_rng(1234)
prices = rng.normal(size=[D])

"""
build cvx problem
"""

# variables
qty_ = cp.Variable(D)  # number of shares bought (qty_>0) or sold (qty<0)
inv_ = cp.Variable(D)  # number of shares in portfolio

# parameters
_lam = cp.Parameter(1, nonneg=True)

# constraints
dynamics = [inv_[1:] == inv_[:-1] + qty_[1:]]  # portfolio dynamics
qty_0 = [qty_[0] == 0.0]  # number of shares traded at time 0
qty_l = [qty_ >= -1.0]  # lower bound of shares traded each time
qty_u = [qty_ <= 1.0]  # upper bound of shares traded each time
inv_0 = [inv_[0] == 0.0]  # number of shares hold at time 0
inv_l = [inv_ >= -10.0]  # lower bound of shares hold each time
inv_u = [inv_ <= 10.0]  # upper bound of shares hold each time
inv_Dm1 = [inv_[D-1] == 0.0]  # number of shares hold the last day

constraints = (dynamics +
               qty_0 +
               qty_l +
               qty_u +
               inv_0 +
               inv_l +
               inv_u +
               inv_Dm1)

# objective

objective = cp.Maximize(-qty_@prices + _lam * (cp.sum(cp.log(qty_+1.) + cp.log(-qty_+1.))))

# The objective is composed of 2 terms:
# 1)  "-qty_@prices" is the PnL
# 2)  "cp.sum(cp.log(qty_+1.) + cp.log(-qty_+1.))" is a log-barier penalization when getting close to 1.0 or -1.0
# The parameter _lam sets the balance between term 1) and 2)

Sproblem = cp.Problem(objective, constraints)

"""
build cvx layer
"""
layer = CvxpyLayer(Sproblem, parameters=[_lam], variables=[qty_])

"""
parameters of the numerical computations
"""
solver = 'ECOS'
mode = 'lsqr'  # 'lsqr', 'lsmr' or 'dense'
lam = torch.tensor([0.1])
dlam = 1e-6
lam_pert = lam + dlam

if (solver == 'ECOS'):
    solver_args = dict(solve_method=solver, mode=mode, feastol=1e-10, abstol=1e-10, reltol=1e-10, max_iters=10000)
elif (solver == 'SCS'):
    solver_args = dict(solve_method=solver, mode=mode, eps=1e-10)
else:
    raise ValueError(f'cannot have solver = {solver}')

"""
calculate the sensitivity of the first trade qty[1] to a change in _lam
"""

# calculation 1) automatic differentiation
lam.requires_grad = True
qty, = layer(lam, solver_args=solver_args)
qty[1].backward()
print('grad =', lam.grad.detach().numpy())

# calculation 2) finite difference
qty_pert, = layer(lam_pert, solver_args=solver_args)
diff = (qty_pert[1] - qty[1])/dlam
print('diff =', diff.detach().numpy())
```

**Comment by AxelBreuer** [comment]
Since cvxpylayers uses diffcp, I have reported the same issue on https://github.com/cvxgrp/diffcp/issues/57 but with a simpler example.

**Comment by AxelBreuer** [comment]
closed by https://github.com/cvxgrp/diffcp/pull/59
